lib/autokey/scripting/engine.py

```diff
@@ -483,44 +470,68 @@ def remove_all_temporary(self, folder=None,
         self.configManager.remove_all_temporary(folder,
                 in_temp_parent)
 
+    def validateArguments(self, folder, name, contents,
+                              abbreviations, hotkey, send_mode, window_filter,
+                              show_in_system_tray, always_prompt, temporary,
+                              replaceExistingHotkey):
+        validateType(folder, "folder", model.Folder)
+        # For when we allow pathlib.Path
+        # validateType(folder, "folder",
+        #         [model.Folder, pathlib.Path])
+        validateType(name, "name", str)
+        validateType(contents, "contents", str)
+        validateAbbreviations(abbreviations)
+        validateHotkey(hotkey)
+        validateType(send_mode, "send_mode", model.SendMode)
+        validateType(window_filter, "window_filter", str)
+        validateType(show_in_system_tray, "show_in_system_tray", bool)
+        validateType(always_prompt, "always_prompt", bool)
+        validateType(temporary, "temporary", bool)
+        validateType(replaceExistingHotkey, "replaceExistingHotkey", bool)
+        # TODO: The validation should be done by some controller functions in the model base classes.
+
+        if folder.temporary and not temporary:
+            raise ValueError("Parameter 'temporary' is False, but parent folder is a temporary one. \
+    Phrases created within temporary folders must themselves be explicitly set temporary")
+
+
+    def check_abbreviation_unique(self, abbreviations):
```

Comment:
You have to add the window filter as an argument here. Multiple model items may have the same abbreviation, if both have distinct * window filters. 

(Side note: I’m using this functionality myself for those GPL headers, which I paste using AutoKey Phrases. The Phrases have the same abbreviation, but different window filters. I use it to paste language-specific headers based on the IDE in use.)

https://github.com/autokey/autokey/blob/f0c69ba81fc082030ce35564e1a52102e20d7be8/lib/autokey/configmanager/configmanager.py#L611-L629

So the abbreviation may be considered "unique" by AutoKey, if the window filters are distinct.

### [*] About distinct window filters, (which is unrelated to the PR):

To determine if two window filters, which are a Python regular expression are distinct, one has to check if the intersection of the regular languages given by the REs is empty. I.e. there is no string, that is matched by both REs. 
This test is infeasible on REs. But they can be converted to DFAs, for which such a test is trivial. 
(This conversion can be done by converting first both REs to [ε-NFAs](https://en.wikipedia.org/wiki/Nondeterministic_finite_automaton#NFA_with_%CE%B5-moves) using https://en.wikipedia.org/wiki/Thompson%27s_construction and then to [DFAs](https://en.wikipedia.org/wiki/Deterministic_finite_automaton) using https://en.wikipedia.org/wiki/Powerset_construction.)

But the last step ε-NFA → DFA  may grow in O(2^n) size. So this test is equally infeasible. But I’ve heard that the intersection test can be implemented directly on ε-NFAs. So if there is a library to convert Python REs to ε-NFAs (pro-tip: there currently isn’t), the test "are window filters distinct" can actually be implemented.

Reply:
Wait, so we can't tell if window filters are distinct? Then how was it done before? (I mostly copied the `unique` code from elsewhere in engine, iirc). Was before incorrect? In which case, do you still need me to correct it?

Reply:
> Wait, so we can't tell if window filters are distinct? 

Yes, in the general case.

> Then how was it done before? (I mostly copied the `unique` code from elsewhere in engine, iirc). 

https://github.com/autokey/autokey/blob/14cbb12acef326b77a4c3964c4ec7172cc2118c5/lib/autokey/model.py#L327-L331

It currently simply checks if the RE patterns are unique. This approach seems to be kind-of sufficient, because normally the REs are written to match literals (WM_CLASS) or simple prefixes or suffixes (window titles). At least, there was never a bug report about it.
You have to try to break it to actually trigger the corner cases here.


> Was before incorrect?

Which “before” do you refer to? To me, that question is ambiguous, so I’ll answer all interpretations I can come up with.
- The uniqueness-check performed by AutoKey is technically incorrect, but works during normal use. Solving this is out of scope for this PR.
- Your code should perform the filter-uniqueness check like the GUI does. 

> In which case, do you still need me to correct it?

Yes, please include the window filter in the uniqueness check. The functionality should match the GUI, even if there are some known-bad corner cases in the validation logic used currently. So just make sure you 



Further notes:
The method above that checks the REs using finite automata will solve it for the general case. But this is more like an academic research topic at the moment. Maybe at some point in time, this can be implemented as a standalone library, which is then used by AutoKey. A function that solves the question "Are these 2 Python REs distinct?" may be useful for other projects, too. 
Another caveat: Python REs have some advanced features that cause them to be actually context-free if used. So this problem may be actually unsolvable, in the mathematical sense, as context-free grammars can’t be checked for being intersection-free.

Reply:
Thanks for that explanation. The explanations about DFAs and the theory is interesting too, is that a specific interest of yours?

Reply:
That’s something i learned in university. I happened to start to like the theoretical bits, because I started to grasp that the reasoning the professors gave were true. So that stuff just stuck.

If you know the theoretical bits, you can argue about the feasibility of features in software based on hard maths. Here, window filters are regular languages, so you can take the toolbox for those and use already done proofs to argue about what can be done, how that can be done and what can not be done.

Reply:
Yeah, I get you. That was the main reason I pushed through learning formal language theory. The famous cartoon about proving NP-equivalence comes to mind.